enclave/asym_keys.c

```diff
@@ -480,188 +480,110 @@ void oe_free_key(
-    }
-    else
-    {
-        /* Copy host -> enclave. */
-        if (!oe_is_outside_enclave(data, data_size))
-            OE_RAISE(OE_INVALID_PARAMETER);
-        out_local = (uint8_t*)malloc(data_size);
-    }
+    if (key_buffer_size_out)
+        *key_info_size_out = 0;
 
-    if (out_local == NULL)
-        OE_RAISE(OE_OUT_OF_MEMORY);
+    if (key_info_size_out)
+        *key_info_size_out = 0;
 
-    OE_CHECK(oe_memcpy_s(out_local, data_size, data, data_size));
-
-    *out = out_local;
-    out_local = NULL;
-    result = OE_OK;
-
-done:
-    if (out_local != NULL)
-    {
-        if (to_host)
-            oe_host_free(out_local);
-        else
-            free(out_local);
-    }
-    return result;
-}
-
-void oe_handle_get_public_key_by_policy(uint64_t arg_in)
-{
-    oe_result_t result = OE_UNEXPECTED;
-    oe_get_public_key_by_policy_args_t* uarg =
-        (oe_get_public_key_by_policy_args_t*)arg_in;
-    oe_get_public_key_by_policy_args_t arg;
-    uint8_t* enclave_user_data = NULL;
-    uint8_t* host_key_info = NULL;
-    uint8_t* host_key_buffer = NULL;
-
-    /* Copy arguments to avoid time of use / time of check. */
-    if (!uarg || !oe_is_outside_enclave(uarg, sizeof(*uarg)))
-        return;
-
-    arg = *uarg;
-    arg.key_buffer = NULL;
-    arg.key_buffer_size = 0;
-    arg.key_info = NULL;
-    arg.key_info_size = 0;
-
-    OE_CHECK(_copy_to_from_host(
-        false,
-        arg.key_params.user_data,
-        arg.key_params.user_data_size,
-        &enclave_user_data));
+    if (!key_buffer_size_out || !key_info_size_out)
+        OE_RAISE(OE_INVALID_PARAMETER);
 
-    arg.key_params.user_data = enclave_user_data;
+    memset(&arg, 0, sizeof(arg));
 
     /* Get the key. */
     OE_CHECK(oe_get_public_key_by_policy(
-        arg.seal_policy,
-        &arg.key_params,
+        (oe_seal_policy_t)seal_policy,
+        key_params,
         &arg.key_buffer,
         &arg.key_buffer_size,
         &arg.key_info,
         &arg.key_info_size));
 
-    /* Copy to host memory. */
-    OE_CHECK(_copy_to_from_host(
-        true, arg.key_info, arg.key_info_size, &host_key_info));
-
-    OE_CHECK(_copy_to_from_host(
-        true, arg.key_buffer, arg.key_buffer_size, &host_key_buffer));
+    *key_buffer_size_out = arg.key_buffer_size;
+    *key_info_size_out = arg.key_info_size;
 
-    /* Success. Just copy to unsafe struct now. */
-    uarg->key_info = host_key_info;
-    uarg->key_info_size = arg.key_info_size;
-    uarg->key_buffer = host_key_buffer;
-    uarg->key_buffer_size = arg.key_buffer_size;
-    host_key_info = NULL;
-    host_key_buffer = NULL;
-    result = OE_OK;
+    if (key_buffer_size < arg.key_buffer_size)
+        OE_RAISE(OE_BUFFER_TOO_SMALL);
 
-done:
-    uarg->result = result;
+    if (key_info_size < arg.key_info_size)
+        OE_RAISE(OE_BUFFER_TOO_SMALL);
 
-    if (enclave_user_data != NULL)
-        free(enclave_user_data);
+    memcpy(key_buffer, arg.key_buffer, arg.key_buffer_size);
```

Comment:
Our functions rarely allocate and return buffers to the caller. oe_get_public_key_by_policy seems to be an exception.
Consider making oe_get_public_key_by_policy take an output buffer to write to rather than allocating and returning outputs.

Reply:
For things that need variable size buffers, functions should allocate internally rather than making the caller call twice (which also has problems if the size needed might change in between).   The Microsoft API guidance is to do allocation internally in such cases.

Reply:
@anakrish: the function **oe_get_public_key_by_policy** is preexisting. There are many places where I could have rewritten the entire chain of requests emanating from the edge routine but this PR makes local changes. I noticed that the arguments structures are not confined to the edge routines. They leak into many other functions. This is not ideal but the current PR meets our security goals.

@dthaler: we are trading that ideal to minimize an attack vector. Any **OCALL** that receives memory becomes an attack target. Our general policy is not to pass memory from the host. Also, returning host memory does not play well with the TrustZone memory model nor with the **virtual enclave** memory model (where the enclave runs in a separate process).